photon-ml/src/main/scala/com/linkedin/photon/ml/util/IOUtils.scala

```diff
@@ -320,12 +329,57 @@ protected[ml] object IOUtils {
       forceOverwrite = true)
   }
 
-  private case class BasicSummaryItems(
-      max: Double,
-      min: Double,
-      mean: Double,
-      normL1: Double,
-      normL2: Double,
-      numNonzeros: Double,
-      variance: Double)
+  /**
+   * Write to a stream while handling exceptions, and closing the stream correctly whether writing to it
+   * succeeded or not.
+   *
+   * NOTE: remember that a Try instance can be understood as a collection, that can have zero
+   * or one element.
+   *
+   * @param outputStream A lambda OutputStream generator
+   * @param op A lambda that executes an operation on its PrintWriter argument
+   * @return Success or Failure. In case of Failure, the Failure contains the exception triggered
+   */
+  def safePrintToStream(outputStream: => OutputStream)(op: PrintWriter => Unit): Try[Unit] = {
+
+    val os = Try(outputStream)
+    val writer = os.map(stream => new PrintWriter(stream))
+    val writeOp = writer.map(op(_))
+    val flush = writer.map(_.flush)
+    val close = os.map(_.close)
+    writeOp.flatMap(_ => flush).flatMap(_ => close)
```

Comment:
Probably need to add some comments explaining this approach and why it's equivalent to try/finally cleanup. With all the vars it's easy to lose track of who's mapping over what.